### PR TITLE
doc: add benchmark doc for new storage pkg

### PR DESCRIPTION
**Pull request by xiang90**
Add a documentation about the memory benchmark for new storage pkg. 

The testing method is still naive. We do not really have any production data yet. But it reveals some good memory usage pattern and matches with our design expectation.

### CODE REVIEW

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                      +-----+
+                      | ... |
+                      |     |
+                      +-----+
+                      | ... |
+                      |     |
+                      ------+
+```
+
+[Page cache memory](pagecache] is managed by the operating system. We will not go into details of this part in this documentation.
```
Comment (yichengq) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: (pagecache)

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+                      ------+
+                      | ... |
+                      |     |
+                      +-----+
+                      | ... |
+                      |     |
+                      +-----+
+                      | ... |
+                      |     |
+                      ------+
```
Comment (yichengq) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: what is this?

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md (hunk 1 above)
Comment (xiang90) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: no. it should be [pagecache]

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
+# Storage Memory Usage Benchmark
+
+The storage(todo: link to design doc about storage) has two components that consume physical memory: the _in-memory index_, and the _page cache_ for on-disk data.
+
+The in-memory index holds all the keys in a [B-tree][btree] data structure, along with pointers to the on-disk data (the values). Each key in the B-tree may contain multiple pointers, pointing to different versions of its values. The theoretical memory consumption of the in-memory index can hence be approximated with the formula:
+
+`N * (c1 + avg_key_size) + N * (avg_versions_of_key) * (c2 + size_of_pointer)`
```
Comment (yichengq) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: what is `size_of_pointer` here? is it fixed 8 bytes in 64-bit machines?

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+| 100K | 5        | 64bytes  | 39MB         |
+| 1M   | 1        | 64bytes  | 218MB        |
+| 1M   | 5        | 64bytes  | 432MB        |
+| 100K | 1        | 256bytes | 41MB         |
+| 100K | 5        | 256bytes | 65MB         |
+| 1M   | 1        | 256bytes | 409MB        |
+| 1M   | 5        | 256bytes | 506MB        |
+
+
+Based on the result, we can calculate `c1=120bytes`, `c2=30bytes`. The key metadata overhead is still relatively nontrivial (50%) for small key-value pairs. However, this is a significant improvement over the old store, which had at least 1000% overhead.
```
Comment (yichengq) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: I like the `c1` and `c2` calculation. :)

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md (hunk 3 above)
Comment (xiang90) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: `along with pointers to the on-disk data`. This pointer is not what you were referring to.

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md (hunk 4 above)
Comment (xiang90) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: I did not show how I calculate it though. It is a magic.

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md (hunk 2 above)
Comment (xiang90) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: see line 11
```
The graph shows the detailed structure of the in memory index B-tree.
```

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md (hunk 3 above)
Comment (yichengq) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: i see. is it possible to provide the estimation range of the `size_of_pointer`?

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                      |     |
+                      ------+
+```
+
+[Page cache memory][pagecache] is managed by the operating system. We will not go into details of this part in this documentation.
+
+## Testing Environment
+
+etcd version
+- git head 776e9fb7be7eee5e6b58ab977c8887b4fe4d48db
```
Comment (jonboulle) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: Could link to https://github.com/coreos/etcd/commit/776e9fb7be7eee5e6b58ab977c8887b4fe4d48db

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md (hunk 4 above)
Comment (jonboulle) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: can we explain that?

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+The storage(todo: link to design doc about storage) has two components that consume physical memory: the _in-memory index_, and the _page cache_ for on-disk data.
+
+The in-memory index holds all the keys in a [B-tree][btree] data structure, along with pointers to the on-disk data (the values). Each key in the B-tree may contain multiple pointers, pointing to different versions of its values. The theoretical memory consumption of the in-memory index can hence be approximated with the formula:
+
+`N * (c1 + avg_key_size) + N * (avg_versions_of_key) * (c2 + size_of_pointer)`
+
+where `c1` is the key metadata overhead and `c2` is the version metadata overhead.
+
+The graph shows the detailed structure of the in memory index B-tree.
```
Comment (jonboulle) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: %s/in memory index/in-memory index/g

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
+# Storage Memory Usage Benchmark
+
+The storage(todo: link to design doc about storage) has two components that consume physical memory: the _in-memory index_, and the _page cache_ for on-disk data.
```
Comment (jonboulle) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: when is this link going in?

Comment (joshix) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: 1. "etcd storage has two components..."
2. Can we either resolve the TODO before merging/publishing, or move it out of the main text and into a line comment or something?

Comment (xiang90) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: @joshix I will change it to a line comment.

Comment (xiang90) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: @jonboulle I am not sure when. :(

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+                      +-----+
+                      | ... | Tree Node
+                      |     |
+                      +-----+
+                      | ... |
+                      |     |
+                      ------+
+```
+
+[Page cache memory][pagecache] is managed by the operating system. We will not go into details of this part in this documentation.
```
Comment (joshix) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: [Page cache memory][pagecache] is managed by the operating system and is not covered in detail in this document.

File: Documentation/benchmarks/etcd-storage-memory-benchmark.md [hunk 9]
```diff
@@ -0,0 +1,98 @@
+# Storage Memory Usage Benchmark
+
+<!---todo: link storage to storage design doc-->
+The storage has two components that consume physical memory: the _in-memory index_, and the _page cache_ for on-disk data.
```
Comment (joshix) [inline Documentation/benchmarks/etcd-storage-memory-benchmark.md]: Two components of etcd storage consume physical memory. The etcd process allocates an *in-memory index* to speed key lookup. The process's *page cache*, managed by the operating system, stores recently-accessed data from disk for quick re-use.

**Pull request merged**